### ISSUE TITLE
Suggestion: a new built-in method to calculate getExtent() for GeoImage

### ISSUE THREAD

**Issue by denilsonsa** [Viglino/ol-ext#330]
**TL;DR:** I wish `GeoImage` source had a built-in method to return its own extent. And I wish the `Layer` object could use such method to automatically return the layer extent.

---

**My use-case:**

1. I have a `GeoImage` source in a `Image` layer. This GeoImage has scaling and rotation, and possibly a crop (or maybe a mask).
2. I am using `LayerSwitcher({ extent: true })` to display a button to auto-zoom to that layer.
    * [LayerSwitcher can only display such button if the layer has extent set.](https://github.com/Viglino/ol-ext/blob/v3.1.4/src/control/LayerSwitcher.js#L714)
3. There is no built-in way to calculate the final geo-image coordinates after all the transformations. (see #116)
4. In order to enable my point 2, I thought about using my point 3 to calculate the extent, and then call `setExtent()` on the image layer. However, 3 does not exist yet.

My current workaround is to copy-paste the code from #116 into my own code. However, I believe it's a common use-case that it should be built into GeoImage itself.

**Suggestion:**

`GeoImage` object should have some method to return the polygon/rectangle/extent/coordinates of the transformed image. This only works after the image has been loaded.

Some options to the method could be:

* Return the value before or after cropping. (I believe the default should be after cropping.)
* Return the values considering or not considering the imageMask. (This is probably redundant, as the imageMask is already defined in the map coordinates, but it still makes sense here, because the caller of this function just wants to get the bounds, regardless if there is a mask or not.)

**Bonus suggestion:**

It would be nice to have a special Layer class that could automatically calculate the extent from its GeoImage source. For instance, I wish something like this would be possible:

```javascript
    var layer = new ol.layer.GeoImage({
        source: new ol.source.GeoImage({ ... })
    });
    
    // Insert some logic here to make sure the source image has been loaded.
    // Alternatively, we could pass previously-known image dimensions
    // to the `ol.source.GeoImage` constructor options.
    
    var extent = layer.getExtent();
    // This extent is valid and automatically calculated from the source GeoImage.
```

I imagine this can be possible by subclassing `ol.layer.Image` into `ol.layer.GeoImage`, and only replacing one method: `getExtent`. This new code would call something like `this.getSource().calculateExtentFromGeoImage()`, instead of only relying on values set by `setExtent`.

**Impact:**

Once this is implemented, developers would be able to:

* Easily draw a polygon on the image. (And use it with the Transform interaction, as desired by #116.)
* Easily calculate if the image is visible on screen. (Something that [LayerSwitcher](https://github.com/Viglino/ol-ext/blob/v3.1.4/src/control/LayerSwitcher.js#L366) tests.)
* Easily pan-and-zoom (and rotate) the map to make the image visible. (In other words, the "zoom to extent" button from LayerSwitcher.)
* If the bonus suggestion is also implemented, then all of this could work out-of-the-box, no extra workaround nor configuration needed. LayerSwitcher would be able to automatically detect visibility and zoom to the GeoImage layers.

---

**Sidenote:** I see some code smell at [`viewChange`](https://github.com/Viglino/ol-ext/blob/v3.1.4/src/control/LayerSwitcher.js#L302), because over there it has a duplicate implementation of  [`testLayerVisibility`](https://github.com/Viglino/ol-ext/blob/v3.1.4/src/control/LayerSwitcher.js#L366).

**Comment by Viglino** [comment]
Good idea.
I'll try to see what can be done asap. :+1:

**Comment by Viglino** [comment]
I've added a `calculateExtent` and `getExtent` method on GeoImage source.
A new layer GeoImage is available: it's a standard Image layer that takes source extent into account.
Look at [example](https://viglino.github.io/ol-ext/examples/layer/map.geoimage.html) to see it working with the layer switcher.
The returned extent doesn't take into account the crop parameter but fit the mask when exists.
A boolean parameter in `calculateExtent` lets you get the extent of the image or the extent of the mask (default use the mask).
The extent is re-calculated on each change (and image load).

**Comment by denilsonsa** [comment]
Awesome!

Question: Why does [`getExtent()`](https://github.com/Viglino/ol-ext/commit/89484b29ee629c2057cecaa5c1fa5fa028b8f16e?w=1#diff-05a600f05b4722ca90360a5ca37843cfR244) receive an extent as a parameter? I understand it is to allow replacing an extent object in-place, but why? The [`layer/Base.js`](https://github.com/openlayers/openlayers/blob/v5.3.0/src/ol/layer/Base.js#L133) doesn't have it. Oh, I see, you're replicating [`source/Vector.js`](https://github.com/openlayers/openlayers/blob/v5.3.0/src/ol/source/Vector.js#L746), which calls [`structs/RBush.js`](https://github.com/openlayers/openlayers/blob/v5.3.0/src/ol/structs/RBush.js#L223). Maybe you can replace that extra logic that you wrote with [`extent.createOrUpdate()`](https://github.com/openlayers/openlayers/blob/v5.3.0/src/ol/extent.js#L215).

If you want to add support for Crop when calculating the extent, I have this piece of code that I wrote:

```js
get_polygon_from_geo_image(geo_image_source) {
	// Inspired by https://github.com/Viglino/ol-ext/issues/116
	var center = geo_image_source.getCenter();
	var scale = geo_image_source.getScale();
	var crop = geo_image_source.getCrop();
	var pixel_width = geo_image_source.getGeoImage().width;
	var pixel_height = geo_image_source.getGeoImage().height;
	var min_x = center[0] + (-pixel_width  / 2 + crop[0]) * scale[0];
	var min_y = center[1] - (-pixel_height / 2 + crop[1]) * scale[1];
	var max_x = center[0] + (-pixel_width  / 2 + crop[2]) * scale[0];
	var max_y = center[1] - (-pixel_height / 2 + crop[3]) * scale[1];
	var extent = ol.extent.boundingExtent([[min_x, min_y], [max_x, max_y]]);
	var polygon = ol.geom.Polygon.fromExtent(extent);
	polygon.rotate(-geo_image_source.getRotation(), center);
	return polygon;
}
```

Feel free to use it. I think it will be a worth addition, given that Mask is already taken into consideration.

Awesome work!

**Comment by denilsonsa** [comment]
> **Sidenote:** I see some code smell at [`viewChange`](https://github.com/Viglino/ol-ext/blob/v3.1.4/src/control/LayerSwitcher.js#L302), because over there it has a duplicate implementation of [`testLayerVisibility`](https://github.com/Viglino/ol-ext/blob/v3.1.4/src/control/LayerSwitcher.js#L366).

FTR: I see this was fixed in https://github.com/Viglino/ol-ext/commit/150e54c1e1d7a126145dde7c5ff5b70619a10a3a?w=1#diff-6b0e2658ff3b18eb2a8e60bc889026d1R301-R322 and cleaned up by ad04934dcab105b2c46e94ee166d3f575a75b00b.

**Comment by Viglino** [comment]
> Question: Why does [`getExtent()`](https://github.com/Viglino/ol-ext/commit/89484b29ee629c2057cecaa5c1fa5fa028b8f16e?w=1#diff-05a600f05b4722ca90360a5ca37843cfR244) receive an extent as a parameter? 

The GeoImage source getExtent mimic the other sources getExtent (Vector sources) with an extent parameter (optional).
The GeoImage layer has no parameter as other layers objects.

**Comment by mdellanave** [comment]
@Viglino @denilsonsa 
Hi guys, first of all thank you, here I found some inputs to solve my problems...

I'm trying to scale an image GeoImage using a VectorLayer with transform interaction active.

here a live demo
https://codesandbox.io/s/adoring-architecture-or1lm?file=/src/index.js

I cannot find a way to "bind" the dimension on the GeoImage after a scale event.
The problem come from the fact that the polygon is redimensioned and the scale factor always restart from 1 after the scaleend event.
Knowing the width and height in pixel of the polygon could solve the problem with some math but I don't know how to get those value.
My idea is to find geoImage scalex = (poly scale x * geoImage with ) / poly with

Any solution?
